prog/dftb+/prg_dftb/parser.F90

```diff
@@ -5142,6 +5145,64 @@ end subroutine finalizeNegf
 #:endif
 
 
+  !> This subroutine overrides the neutral (reference) atom electronic occupation 
+  subroutine readCustomReferenceOcc(root, orb, referenceOcc, geo, iAtInRegion, &
+      & customOcc)
+    type(fnode), pointer, intent(in) :: root
+    type(TOrbitals), intent(in) :: orb
+    real(dp), intent(in) :: referenceOcc(:,:)
+    type(TGeometry), intent(in) :: geo
+    type(WrappedInt1), allocatable, intent(out) :: iAtInRegion(:)
+    real(dp), allocatable, intent(out) :: customOcc(:,:)
+
+    type(fnode), pointer :: node, container, child
+    type(fNodeList), pointer :: nodes
+    type(string) :: buffer
+    integer :: nCustomOcc, iCustomOcc, iShell, iSpecies, nAtom
+    character(sc) :: shellname
+    logical, allocatable :: atomOverriden(:)
+
+    call getChild(root, "CustomReferenceOccupations", container, requested=.false.)
+    if (.not. associated(container)) then
+      return
+    end if
+
+    call getChildren(container, "ReferenceOccupation", nodes)
+    nCustomOcc = getLength(nodes)
+    nAtom = size(geo%species)
+    allocate(iAtInRegion(nCustomOcc))
+    allocate(customOcc(orb%mShell, nCustomOcc))
+    allocate(atomOverriden(nAtom))
+    atomOverriden(:) = .false.
+    customOcc(:,:) = 0.0_dp
+    
+    do iCustomOcc = 1, nCustomOcc
+      call getItem1(nodes, iCustomOcc, node)
+      call getChildValue(node, "Atoms", buffer, child=child, multiple=.true.)
+      call convAtomRangeToInt(char(buffer), geo%speciesNames, &
+          & geo%species, child, iAtInRegion(iCustomOcc)%data)
+      if (any(atomOverriden(iAtInRegion(iCustomOcc)%data))) then
+        call detailedError(child, "Atom region contains atom(s) which have&
+            & already been overriden")
+      end if
+      atomOverriden(iAtInRegion(iCustomOcc)%data) = .true.
+      iSpecies = geo%species(iAtInRegion(iCustomOcc)%data(1))
+      if (any(geo%species(iAtInRegion(iCustomOcc)%data) /= iSpecies)) then
+        call detailedError(child, "All atoms in a ReferenceOccupation&
+            & declaration must have the same type.")
+      end if
+      do iShell = 1, orb%nShell(iSpecies)
+        write(shellname, "(A)") orbitalNames(orb%angShell(iShell, iSpecies) + 1)
+          call getChildValue(node, shellname, customOcc(iShell, iCustomOcc), &
```

Comment:
Not compatible with the extended basis set structure in the code (chance of multiple s shells for example).

Reply:
Then I don't know the logic you have in mind to treat extra shells. The orbital names:

orbitalNames(orb%angShell(iShell, iSpecies) + 1)

shouldn't be different for different shells ? So for an extra s they can have a name like s2 (or s* d*, as I suggested). Otherwise how should I differentiate ? The way it was done in the svn code was also wrong in my opinion.

Reply:
Ok, all the point is to create a unique 'label' for shellname that can be searched for in the input.
I can write a subroutine that create that string with the following convention:
First non-repeated shells with normal names 's','p',... then as soon as an OrbitalName repeats it is assigned a name like s2, s3, d2, ...  
This is just for the string to match. So user can use normal 's' 'p' 'd' names in usual calculations. 

Side Question: How do you set MaxAngularMomentum in case of sps* ?